# Домашнее задание № 6

## по дисциплине: Теория вероятностей и математическая статистика

## Тема: Взаимосвязь величин. Показатели корреляции. Корреляционный анализ. Проверка на нормальность

### выполнил: Евгений Кириллов

__Задача 1__

Даны значения величины заработной платы заемщиков банка (zp) и значения их
поведенческого кредитного скоринга (ks):

zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],

ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].

Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с
помощью функции cov из numpy.
Полученные значения должны быть равны.
Найдите коэффициент корреляции Пирсона с помощью ковариации и
среднеквадратичных отклонений двух признаков,
а затем с использованием функций из библиотек numpy и pandas.

In [1]:
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110]
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]

найдём ковариацию с помощью элементарных действий:

In [2]:
zp_mean = sum(zp) / len(zp)
ks_mean = sum(ks) / len(ks)

zp_mean, ks_mean

(101.4, 709.9)

In [3]:
zp_ks_cov = sum([(i[0] - zp_mean) * (i[1] - ks_mean) for i in zip(zp, ks)]) / (len(zp) - 1)
zp_ks_cov

10175.377777777778

найдём ковариацию с помощью функции cov из numpy:

In [4]:
import numpy as np

In [5]:
np.cov(zp, ks, ddof=1)

array([[ 3882.93333333, 10175.37777778],
       [10175.37777778, 33854.32222222]])

полученные значения равны

найдём коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков:

In [6]:
zp_std = (sum([(i - zp_mean) ** 2 for i in zp]) / (len(zp) - 1)) ** 0.5
ks_std = (sum([(i - ks_mean) ** 2 for i in ks]) / (len(ks) - 1)) ** 0.5

zp_std, ks_std

(62.31318747531162, 183.99544076477068)

In [7]:
zp_ks_corr = zp_ks_cov / (zp_std * ks_std)
zp_ks_corr

0.8874900920739162

найдём коэффициент корреляции Пирсона с помощью функции np.corrcoef() из библиотеки numpy:

In [8]:
np.corrcoef(zp, ks)

array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

найдём коэффициент корреляции Пирсона с помощью метода .corr() из библиотеки pandas:

In [9]:
import pandas as pd

In [10]:
pd.DataFrame({'zp': zp, 'ks':ks}).corr()

zp       ks
zp  1.00000  0.88749
ks  0.88749  1.00000

полученные значения равны

__Задача 2__

Проведите тест на значимость коэффициента корреляции Пирсона, найденного в предыдущей задаче. Что для этого нужно знать:
* Нулевая гипотеза: реальный коэффициент корреляции равен 0. Альтернативная гипотеза двухсторонняя.
* Статистика: `t = r * sqrt(n - 2) / sqrt(1 - r ** 2)`, где `r` - коэффициент корреляции Пирсона, посчитанный по выборке.
* В предположении верности нулевой гипотезы эта статистика имеет распределение Стьюдента с параметром `df = n - 2`.

В качестве ответа запишите 1, если нулевая гипотеза верна, и 0 иначе.

In [11]:
from scipy import stats

In [12]:
alpha = 0.05
n = len(zp)

найдём квантили:

In [13]:
t1 = stats.t.ppf(alpha / 2, df=n - 2)
t2 = stats.t.ppf(1 - alpha / 2, df=n - 2)

t1, t2

(-2.306004135033371, 2.3060041350333704)

найдём значение статистики t:

In [14]:
r = zp_ks_corr
t = r * np.sqrt(n - 2) / np.sqrt(1 - r ** 2)
t

5.447168150485575

5.447168150485575 > 2.3060041350333704

статистика попала в критическую область, нулевая гипотеза отвергается

__Задача 3__ 

Измерены значения IQ выборки студентов,
обучающихся в местных технических вузах:

131, 125, 115, 122, 131, 115, 107, 99, 125, 111.

Известно, что в генеральной совокупности IQ распределен нормально.
Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [15]:
from scipy import stats

In [16]:
iq = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])

In [17]:
n = iq.shape[0]
mean = iq.mean()
std = iq.std(ddof=1)
p = 0.95
alpha = 1 - p

найдём границы критической области:

т.к. дисперсия генеральной совокупности неизвестна, то используем t-статистику

In [18]:
t1 = stats.t.ppf(alpha / 2, df=n - 1)
t2 = stats.t.ppf(1 - alpha / 2, df=n - 1)

t1, t2

(-2.2621571627409915, 2.2621571627409915)

найдём границы доверительного интервала:

In [19]:
(mean + t1 * std / np.sqrt(n), mean + t2 * std / np.sqrt(n))

(110.55608365158724, 125.64391634841274)

и его ширину:

In [20]:
(mean + t2 * std / np.sqrt(n)) - (mean + t1 * std / np.sqrt(n))

15.087832696825501

__Задача 4__

Известно, что рост футболистов в сборной распределен нормально
с дисперсией генеральной совокупности, равной 25 кв.см. Объем выборки равен 27,
среднее выборочное составляет 174.2. Найдите доверительный интервал для
математического ожидания с надежностью 0.95.

In [21]:
n = 27
mean = 174.2
var = 25
std = np.sqrt(var)
p = 0.95
alpha = 1 - p

найдём границы критической области:

т.к. дисперсия генеральной совокупности известна, то используем Z-статистику

In [22]:
t1 = stats.norm.ppf(alpha / 2)
t2 = stats.norm.ppf(1 - alpha / 2)

t1, t2

(-1.959963984540054, 1.959963984540054)

найдём границы доверительного интервала:

In [23]:
(mean + t1 * std / np.sqrt(n), mean + t2 * std / np.sqrt(n))

(172.3140237765397, 176.08597622346028)

и его ширину:

In [24]:
(mean + t2 * std / np.sqrt(n)) - (mean + t1 * std / np.sqrt(n))

3.7719524469205794